### PR TITLE
Implement a Client to retrieve tenant branding preferences

### PR DESCRIPTION
**Pull request by brionmario**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

- Added a new API Client `BrandingPreferenceRetrievalClient` to interact with the `/api/server/v1/branding-preference` API.(https://github.com/wso2/carbon-identity-framework/pull/3831)

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

- Should only be merged after https://github.com/wso2/carbon-identity-framework/pull/3831 is merged.

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

- Bump the framework version in identity-apps.

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

**Review by inthirakumaaran — APPROVED** [review]

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+        try (CloseableHttpClient httpclient = HttpClientBuilder.create().useSystemProperties().build()) {
+
+            String uri = getBrandingPreferenceEndpoint(tenant);
+
+            try {
+                URIBuilder uriBuilder = new URIBuilder(uri);
+
+                if (StringUtils.isNotBlank(type)) {
+                    if (log.isDebugEnabled()) {
```
Comment (inthirakumaaran) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java:75]: Shall we improve the debug log a bit? we can try to add tenant, combine all the params in a single debug log WDYT?

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+                        log.debug("Appending resource type to the preferences endpoint URI : " + type);
+                    }
+
+                    uriBuilder.addParameter(RESOURCE_TYPE_URL_SEARCH_PARAM, type);
+                }
+
```
Comment (inthirakumaaran) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java:81]:
```suggestion
```

Comment (inthirakumaaran) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java:81]: Remove unnecessary space as much as possible, if needed add spaces between different logic blocks.

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+                    uriBuilder.addParameter(RESOURCE_LOCALE_URL_SEARCH_PARAM, locale);
+                }
+            } catch (URISyntaxException e) {
+                String msg = "Error while building the branding preference endpoint URI with params for : "
+                        + tenant + ". Falling back to default endpoint URI: " + uri;
+
+                if (log.isDebugEnabled()) {
+                    log.debug(msg, e);
+                }
```
Comment (inthirakumaaran) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java:103]:
```suggestion
                if (log.isDebugEnabled()) {
                    String msg = "Error while building the branding preference endpoint URI with params for : "
                        + tenant + ". Falling back to default endpoint URI: " + uri;
                    log.debug(msg, e);
                }
```

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClientException.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+     */
+    public BrandingPreferenceRetrievalClientException(String message, Throwable throwable) {
+
+        super(message, throwable);
+    }
+
+    /**
+     * Client Exception with error message.
+     *
+     * @param message Error message
```
Comment (inthirakumaaran) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClientException.java:37]:
```suggestion
     * @param message Error message.
```

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+    /**
+     * Check for branding preference in the given tenant.
+     *
+     * @param tenant Tenant Domain.
+     * @param type Resource Type. ex: ORG, APP, etc.
+     * @param name Resource Name. ex: Console, My Account, etc.
+     * @param locale ISO language code of the resource. ex: en-US, pt-BR, etc.
+     * @return A JSON Object containing branding preferences.
+     * @throws BrandingPreferenceRetrievalClientException
```
Comment (inthirakumaaran) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java:62]:
```suggestion
     * @throws BrandingPreferenceRetrievalClientException.
```

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClientException.java (hunk 4 above)
Comment (brionmario) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClientException.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/3831/commits/e17df8230b25ece4ab75d5372a3464beeecb2543

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java (hunk 3 above)
Comment (brionmario) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/3831/commits/e17df8230b25ece4ab75d5372a3464beeecb2543

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java (hunk 2 above)
Comment (brionmario) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java:77]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/3831/commits/e17df8230b25ece4ab75d5372a3464beeecb2543

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java (hunk 1 above)
Comment (brionmario) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/3831/commits/e17df8230b25ece4ab75d5372a3464beeecb2543

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+import java.nio.charset.Charset;
+
+/**
+ * Client to interact with the Tenant branding preferences API.
+ */
+public class BrandingPreferenceRetrievalClient {
+
+    private static final String CLIENT = "Client ";
+    private static final Logger log = Logger.getLogger(PreferenceRetrievalClient.class);
+    private static final String BRANDING_PREFERENCE_API_RELATIVE_PATH = "/api/server/v1/branding-preference";
```
Comment (brionmario) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java:49]: @UdeshAthukorala Please validate.

Comment (UdeshAthukorala) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint.util/src/main/java/org/wso2/carbon/identity/mgt/endpoint/util/client/BrandingPreferenceRetrievalClient.java:49]: Ack. This is the endpoint.

**Pull request merged**